code/python/create_template_from_Syn_schema.py

```diff
@@ -0,0 +1,270 @@
+        values_df - A pandas dataframe of the values lists used by columns
+                    in the template
+    """
+    base_file_name, base_file_ext = os.path.splitext(template_file_name)
+    dictionary_file_name = (base_file_name + "_dictionary" +
+                            base_file_ext)
+    values_file_name = (base_file_name + "_values" + base_file_ext)
+
+    # Create a template file.
+    with open(template_file_name, "w") as template_file:
+        template_df.to_csv(template_file, index=False)
+
+    # Create a dictionary file
+    with open(dictionary_file_name, "w") as dictionary_file:
+        dictionary_df.to_csv(dictionary_file, index=False)
+
+    # Create a values file
+    with open(values_file_name, "w") as values_file:
+        values_df.to_csv(values_file, index=False)
+
+
+def template_excel(workbook_name, template_df, dictionary_df, values_df):
+    """
+    Function: template_excel
+
+    Purpose: Write an Excel workbook containing the following worksheets:
+             - "Template": worksheet with column names to be used for
+                           entering data
+             - "Dictionary": worksheet with the definitions of the columns in
+                             the "Template" worksheet
+             - "Values": worksheet with any controlled vocabulary lists used
+                         for columns in the "Template" worksheet
+
+    Arguments:
+        workbook_name: Name of the workbook to output
+        template_df - An empty pandas dataframe with the column names being
+                      the desired template columns
+        dictionary_df - A pandas dataframe containing the definition
+                        information for each column in the template
+        values_df - A pandas dataframe of the values lists used by columns
+                    in the template
+    """
+    workbook_writer = pd.ExcelWriter(workbook_name, engine="xlsxwriter")
+
+    # Create a template worksheet.
+    template_df.to_excel(workbook_writer, index=False, sheet_name="Template")
+
+    # Create a dictionary worksheet
+    dictionary_df.to_excel(workbook_writer, index=False, sheet_name="Dictionary")
+
+    # Create a values worksheet
+    values_df.to_excel(workbook_writer, index=False, sheet_name="Values")
+
+    workbook_writer.save()
+
+
+def main():
+
+    syn = synapseclient.Synapse()
+    syn.login(silent=True)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("json_schema_name", type=str,
+                        help="Full name for the registered JSON schema")
+    parser.add_argument("output_file", type=str,
+                        help="Full pathname for the output file")
+    parser.add_argument("type_of_output", type=str,
+                        help="Type of output (csv or excel)")
+
+    args = parser.parse_args()
+
+    json_schema = syn._waitForAsync("/schema/type/validation/async", {"$id": args.json_schema_name})
```

Comment:
This is neat. I generally call the async API, then do a second call with the token returned to get the dereferenced schema. Is this doing the same thing, but with only one function?

Reply:
This function makes the call, gets the token, and then loops in the client with an exponential backoff until the job is finished and returns the result. It is originally meant for internal use not really as a public API, but I know it has become used.

Reply:
@Aryllen @jkiang13 I got this from Bruce, but am willing to change it if it's not appropriate for this use.